pkg/certificate/authorization/endpoint.go

```diff
@@ -0,0 +1,168 @@
+}
+
+// NewEndpoint creates a new authorization gRPC server.
+func NewEndpoint(log *zap.Logger, db *DB, listener net.Listener) *Endpoint {
+	mux := http.NewServeMux()
+	endpoint := &Endpoint{
+		log:      log,
+		db:       db,
+		listener: listener,
+		server: http.Server{
+			Handler: mux,
+		},
+	}
+
+	mux.HandleFunc("/v1/authorization/create", endpoint.httpCreate)
+
+	return endpoint
+}
+
+// Create creates an authorization from the given authorization request.
+func (endpoint *Endpoint) Create(ctx context.Context, req *pb.AuthorizationRequest) (_ *pb.AuthorizationResponse, err error) {
+	mon.Task()(&ctx, req.UserId)(&err)
+
+	existingGroup, err := endpoint.db.Get(ctx, req.UserId)
+	if err != nil {
+		msg := "error getting authorizations"
+		err := ErrEndpoint.New(msg)
+		endpoint.log.Error(msg, zap.Error(err))
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	if len(existingGroup) > 0 {
+		authorization := existingGroup[0]
+		return &pb.AuthorizationResponse{
+			Token: authorization.Token.String(),
+		}, nil
+	}
+
+	createdGroup, err := endpoint.db.Create(ctx, req.UserId, 1)
+	if err != nil {
+		msg := "error creating authorization"
+		err := ErrEndpoint.New(msg)
+		endpoint.log.Error(msg, zap.Error(err))
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	groupLen := len(createdGroup)
+	if groupLen != 1 {
+		clientMsg := "error creating authorization"
+		internalMsg := clientMsg + fmt.Sprintf("; expected 1, got %d", groupLen)
+
+		endpoint.log.Error(internalMsg)
+		return nil, status.Error(codes.Internal, ErrEndpoint.New(clientMsg).Error())
+	}
+
+	authorization := createdGroup[0]
+
+	return &pb.AuthorizationResponse{
+		Token: authorization.Token.String(),
+	}, nil
+}
+
+// Run starts the endpoint HTTP server and waits for the context to be
+// cancelled or for `Close` to be called.
+func (endpoint *Endpoint) Run(ctx context.Context) (err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	ctx, cancel := context.WithCancel(ctx)
+	var group errgroup.Group
+	group.Go(func() error {
+		<-ctx.Done()
+		return endpoint.server.Shutdown(context.Background())
+	})
+	group.Go(func() error {
+		defer cancel()
+		return endpoint.server.Serve(endpoint.listener)
+	})
+
+	return group.Wait()
+}
+
+// Close closes the endpoint HTTP server.
+func (endpoint *Endpoint) Close() error {
+	return endpoint.server.Close()
+}
+
+func (endpoint *Endpoint) httpCreate(writer http.ResponseWriter, httpReq *http.Request) {
+	var err error
+	ctx := context.Background()
+	mon.Task()(&ctx)(&err)
+
+	if httpReq.Method != http.MethodPost {
+		msg := fmt.Sprintf("unsupported HTTP method: %s", httpReq.Method)
+		err = ErrEndpoint.New(msg)
+		endpoint.log.Error(msg, zap.Error(ErrEndpoint.New(msg)))
+		http.Error(writer, msg, http.StatusBadRequest)
+		return
+	}
+
+	userID, err := ioutil.ReadAll(httpReq.Body)
+	if bytes.Equal([]byte{}, userID) {
+		msg := "missing user ID body"
+		err = ErrEndpoint.Wrap(err)
+		endpoint.log.Error(msg, zap.Error(err))
```

Comment:
nit: I don't think this needs to be logged. The message is clear what needs to be added to fix request.

Reply:
We should only log system errors, user input errors shouldn't be logged because they don't provide us any information about our system isn't behaving correctly.

Reply:
:+1: that makes sense